iothub/service/src/Messaging/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Azure.Devices.Common.Exceptions;
+using Microsoft.Azure.Devices.Common.Extensions;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Subclient of <see cref="IotHubServiceClient"/> for handling file upload notications.
+    /// </summary>
+    /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload#service-file-upload-notifications"/>.
+    public class FileUploadNotificationProcessorClient : IDisposable
+    {
+        private readonly string _hostName;
+        private readonly IotHubConnectionProperties _credentialProvider;
+        private readonly IotHubConnection _connection;
+
+        /// <summary>
+        /// The callback to be executed each time file upload notification is received from the service.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Func<FileNotification, DeliveryAcknowledgement> _fileNotificationProcessor;
+
+        /// <summary>
+        /// The callback to be executed when the connection is lost.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Action<Exception> _errorProcessor;
```

Comment:
Don't have this action give the user the Exception like this. Wrap it into a class like we did in Java [here](https://github.com/Azure/azure-iot-sdk-java/blob/c35f7782e400d3d7cc341b370117e70ece76236f/service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/messaging/FileUploadNotificationProcessorClient.java#L43). The reason we do this is so that we can add additional fields later such as DateTime when it happened, previous state, etc. without causing breaking changes